Bump to v0.4.0

cr1901 opened the pull request:
cc: @YuhanLiin Do you have any feedback/objections?

cr1901 committed:
Switch from bare-metal to critical-section. Bump to 0.3.1. cargo fmt.

cr1901 committed:
Use u16 for RawRestoreState- generates better code.

cr1901 committed:
Rewrite interrupt::free() to resemble critical_section::with(). Use inline attribute to avoid rustc misoptimizing either case.

cr1901 committed:
Create features for outlining critical sections. Change critical-section feature name to not confict with crate name.

cr1901 committed:
Add new features to README.md.

cr1901 committed:
It has become clear that the critical-section-single-core feature is a breaking change. Bump to 0.4.0.

cr1901 committed:
Update CHANGELOG.md.

cr1901 commented:
I've tested this locally, and the crate appears to work fine. If necessary, we can do a point release. But I need to get these crates released for now. So I'm temporarily disabling the protection on the branch.

YuhanLiin commented on src/interrupt.rs:
```diff
@@ -39,23 +39,38 @@ pub unsafe fn enable() {
     let cs = unsafe { CriticalSection::new() };
     let r = f(&cs);
 
     // If the interrupts were active before our `disable` call, then re-enable
     // them. Otherwise, keep them disabled
```
Delete this comment

cr1901 replied:
Why does this comment need to be deleted? In the context of the deprecated `free` function, I believe the comment is accurate/correct (`free` is hardcoded to disable interrupts).

I've fixed the other things locally.

YuhanLiin replied:
The logic of checking the status to decide whether to re-enable the interrupt is inside the `release` function, so this comment should be moved there.

Also, why isn't `acquire` used in this call?

YuhanLiin commented on src/register/sr.rs:
```diff
@@ -4,6 +4,7 @@ use crate::asm;
 
 /// Status Register
 #[derive(Clone, Copy, Debug)]
+#[repr(C)]
```
If we want `Sr` to have the exact layout as `u16`, then we should use `repr(transparent)` instead.

YuhanLiin commented on src/critical_section.rs:
```diff
@@ -0,0 +1,44 @@
+        // I don't believe #[inline] or #[inline(always)] hints work for
+        // functions marked as "extern" in another crate (which is the case
+        // for the acquire()/release() here- see critical_section crate).
+        #[cfg_attr(any(feature = "outline-cs", feature = "outline-cs-acq"), inline(never))]
```
You don't need to check for `outline-cs` here, because `outline-cs` already implies `outline-cs-acq` and `outline-cs-rel`, according to Cargo.toml.

cr1901 commented:
@YuhanLiin I think I addressed all comments. If you think everything looks good, I can release `v0.4.1`: https://github.com/rust-embedded/msp430/compare/v0.4.0...HEAD

YuhanLiin commented:
Everything's good for release
